openasip/src/applibs/LLVMBackend/RISCVTDGen.cc

```diff
@@ -0,0 +1,396 @@
+/*
+ Copyright (C) 2024 Tampere University.
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2.1 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+/**
+ * @file RISCVTDGen.cc
+ *
+ * Implementation of RISCVTDGen class.
+ *
+ * @author Kari Hepola 2024 (kari.hepola@tuni.fi)
+ * @note rating: red
+ */
+
+#include "RISCVTDGen.hh"
+
+#include "BEMGenerator.hh"
+#include "BinaryEncoding.hh"
+#include "Exception.hh"
+#include "InstructionFormat.hh"
+#include "Machine.hh"
+#include "MapTools.hh"
+#include "Operation.hh"
+#include "OperationPool.hh"
+#include "OperationTriggeredEncoding.hh"
+#include "OperationTriggeredField.hh"
+#include "OperationTriggeredFormat.hh"
+#include "RISCVFields.hh"
+#include "RISCVTools.hh"
+#include "TCEString.hh"
+#include "Application.hh"
+#include "tce_config.h"
+#include <iostream>
+#include <regex>
+#include <sstream>
+#include <iomanip>
+
+#define DEBUG_RISCV_TDGEN 0
+
+RISCVTDGen::RISCVTDGen(const TTAMachine::Machine& mach)
+    : TDGen(mach, false), bem_(NULL) {
+    bem_ = BEMGenerator(mach).generate();
+    assert(bem_ != NULL);
+    findCustomOps();
+    initializeBackendContents();
+}
+
+std::string
+RISCVTDGen::intToHexString(int num) const {
+    std::stringstream ss;
+    ss << "0x" << std::hex << std::uppercase << num;
+    return ss.str();
+}
+
+std::string
+RISCVTDGen::unsignedToHexString(unsigned num) const {
+    std::stringstream ss;
+    ss << "0x" << std::hex << std::uppercase << num;
+    return ss.str();
+}
+
+// TODO: OpenASIP converts hex numbers to unsigned by default. This converted
+// number might not fit into i32
+std::string
+RISCVTDGen::decimalsToHex(const std::string& pattern) const {
```

Comment:
Same. Belongs to StringTools.

Reply:
I would keep this here, since it is highly specific to the TDGen pattern that it operates on